File: code/python/shortestToChar.py
class Solution(object):
    def shortestToChar(self, S, C):
        """
        :type S: str
        :type C: str
        :rtype: List[int]
        """
        res = [10001 for i in range(len(S))]
        tmp = []
        for i in range(len(S)):
            if S[i] == C:
                tmp.append(i)
        
        pre = -10000
        fore = tmp[0]
        j = 1
        i = 0
        while i < len(S):
            if i <= fore:
                res[i] = min(i-pre,fore-i)
                i = i + 1
            else:
                pre = fore
                if j < len(tmp):
                    fore = tmp[j]
                else:
                    fore = 20000
                j = j + 1
        return res
                

File: code/java/shortestToChar.java
class Solution {
    public int[] shortestToChar(String S, char C) {
        int[] res = new int[S.length()];
        Queue<Integer> tmp = new LinkedList<>();
        for(int i=0;i<S.length();i++){
            if(S.charAt(i) == C)
                tmp.add(i);
        }
        int pre = -10000;
        int fore = tmp.poll();
        
        int i = 0;
        while(i < S.length()){
            if (i <= fore){
                res[i] = Math.min(i-pre,fore-i);
                i++;
            }
            else{
                    pre = fore;
                    if(tmp.isEmpty() == false)
                        fore = tmp.poll();
                    else
                        fore = 20000;
                }
            }
        return res;
    }
}
